### PR TITLE
Improving alpha diversity calculation

### PR DESCRIPTION
**Pull request by qiyunzhu**
@ebolyen This addresses the relevant issue #2014 . Now the alpha diversity metrics return `np.nan` instead of zero when the community is empty.

Exceptions are metrics that are solely based on observed counts, including `sobs`, `singles`, `doubles` and `osd`. They will return zero.

One relevant question is how to deal with one-taxon communities when calculating species evenness metrics. At present, `pielou_e` and `heip_e` return `np.nan`, whereas `simpson_e` returns 1. I tend to think 1 makes sense. Evenness should range between 0 and 1, in which 1 means all species have the same abundance (i.e., completely even). When there is only one species, it is guaranteed that evenness is maximized. On the other hand, it will not be ideal if a non-empty community returns a NaN value. What do you think?

@wasade This also improves how counts are validated. Specifically, it will remove zeros from the counts before calculating alpha diversity. This will save troubles in the downstream implementation of alpha diversity metrics. This will also permit efficiency improvement by directly feeding sparse matrices from BIOM into the calculation. However, the latter is not done yet. This needs further discussion. That being said, the current PR should be safe to use before the final solution.

Also cc'ing @mortonjt 

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [x] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

**Comment by qiyunzhu** [comment]
Hello @ebolyen would you like to comment on whether the consideration of edge cases (empty community, one-species community) is sane in this PR (see my comments above)? Also cc'ing @gregcaporaso

**Comment by qiyunzhu** [comment]
Now there is a significant enrichment of the parameters and documentation of alpha diversity metrics.

**Comment by qiyunzhu** [comment]
@wasade @mortonjt Here is a very large PR. I did some substantial amount of literature review in order to improve the calculation and documentation of all alpha diversity metrics. I wonder if either of you can do a quick sanity check of the technical side of my code (especially counts vector validation and transformation), while trusting my professionality in community ecology?

File: skbio/diversity/tests/test_driver.py [hunk 1]
```diff
@@ -168,20 +168,17 @@ def test_invalid_input_phylogenetic(self):
 
     def test_empty(self):
         # empty vector
-        actual = alpha_diversity('sobs',
-                                 np.array([], dtype=np.int64))
+        actual = alpha_diversity('sobs', np.array([], dtype=np.int64))
         expected = pd.Series([0], dtype=int)
```
Comment (mataton) [inline skbio/diversity/tests/test_driver.py:172]: `expected` should have `dtype=np.int64` to pass on Windows.

This is also true for lines 208 in `test_empty`, line 213 in `test_single_count_vector`, lines 255 and 262 in `test_sobs`, and line 315 in `test_no_ids` in this class.

File: skbio/diversity/tests/test_driver.py [hunk 2]
```diff
@@ -168,20 +168,17 @@ def test_invalid_input_phylogenetic(self):
...
         assert_series_almost_equal(actual, expected)
 
         # array of empty vector
-        actual = alpha_diversity('sobs',
-                                 np.array([[]], dtype=np.int64))
+        actual = alpha_diversity('sobs', np.array([[]], dtype=np.int64))
         expected = pd.Series([0], dtype=int)
```
Comment (mataton) [inline skbio/diversity/tests/test_driver.py:177]: `dtype=np.int64`

File: skbio/diversity/tests/test_driver.py [hunk 3]
```diff
@@ -168,20 +168,17 @@ def test_invalid_input_phylogenetic(self):
...
         assert_series_almost_equal(actual, expected)
 
         # array of empty vectors
-        actual = alpha_diversity('sobs',
-                                 np.array([[], []], dtype=np.int64))
+        actual = alpha_diversity('sobs', np.array([[], []], dtype=np.int64))
         expected = pd.Series([0, 0], dtype=int)
```
Comment (mataton) [inline skbio/diversity/tests/test_driver.py:182]: `dtype=np.int64`.

**Review by mataton — CHANGES_REQUESTED** [review]
Other than the windows tests failures which I commented on, everything looks good.

**Comment by qiyunzhu** [comment]
@mataton Thanks! I have made the edits you suggested.

**Review by mataton — APPROVED** [review]

**Comment by qiyunzhu** [comment]
Let me merge this PR and move on.

**Pull request merged**